### ISSUE TITLE
Headless with Static IP address

### ISSUE THREAD

**Issue by asteele4** [davidmytton/alpine-linux-headless-raspberrypi#3]
Hi David,

Thanks for your package and updated notes to the wiki page.

Whilst I have many years experience with Linux, I'm new to using Alpine Linux and particularly on a RPi4.  I used your package to bootstrap setting up my RPi for headless operation.

The issue I ran into was that once I had logged in, and configured it how I wanted, which included setting a static IP address, I was finding my Pi would keep reconfiguring the interfaces file back to using dhcp!  Being new to Alpine Linux I figured it was some setting I had overlooked.  After many frustrating hours I finally thought to check what the headless customisation was doing and found the problem.

From my reading of the headless.start code it seems it only needs to be run on the first boot to enable one to ssh in the first time.  After that I'd expect the user would make changes they want and save them so headless.start is no longer needed.  But as I said, I'm new to the Alpine way of doing things so this may still be needed when running diskless for example.

Due to my lack of knowledge of the Alpine ecosystem, I'm not sure what the correct fix is.

Some ideas that come to mind are:

1. Simply update the README to mention that once logged in, disable/remove headless.start
2. Add a comment to /etc/network/interfaces noting it will be overwritten on reboot
3. A variation on this would be to use a comment flag that headless.start could detect and when present/absent not override the file.
4. Append to /etc/motd something like "Now you're in, disable/delete headless.start"
5. Some other Alpine linux correct way of only running this once.

Another lesser way this presented itself itself is in changes I'd made to sshd_cofig, but looking at the way that gets modified, it would be less of a problem.  For my use case, I didn't do any testing of the wifi setup.

I hope this is just one of those cases that wasn't considered and now identified there's an obvious and correct fix.

Andrew

**Comment by davidmytton** [comment]
Is this because you didn't run `setup-alpine` after booting?

**Comment by asteele4** [comment]
I did run setup-alpine after booting (many times trying to get it to work.
:-) ).  Seems headless wasn't being disabled.  So would re-run and wipe the
config.  What should be disabling headless.start after its run the first
time?  I could see anything obvious in the script itself that does it.  Is
a side effect of running setup-alpine supposed to disable what's in
/etc/local.d from running?

Thanks

**Comment by sodface** [comment]
@asteele4 

If you look at a previous version of the Alpine wiki entry for this:
https://wiki.alpinelinux.org/w/index.php?title=Raspberry_Pi_-_Headless_Installation&oldid=18840

At the bottom:

```
lbu commit -d
reboot
```
See if that helps?

**Comment by asteele4** [comment]
@sodface

I'm not running it diskless.  I presume that means I lbu doesn't work/isn't
needed.

However, your link was very helpful.  It's the step above what you quoted
which has been removed from the current version of the page and still
required:

Remove the local script service from the default runlevel and delete the
headless setup script so it's not saved when the new overlay is created
later:

rc-update del local default rm /etc/local.d/headless.start

**Comment by asteele4** [comment]
@davidmytton firstly a clarification from your post:

>  ... original instructions with the lbu commit -d and reboot after deleting the headless.start file.

It wasn't the removal of the "lbu commit -d" and "reboot" that was a problem.  Rather the step above that.  The step to remove the "headless.start" file is no longer present.  That's the step I needed to do.  I wasn't sure from that quoted bit of your reply, if you meant the not including the "lbu and reboot" steps or also headless step because they were no longer needed.

Anyway, putting that aside, I just tested it again:

My test setup is booting off a USB stick containing a standard unpacked install tarball, the headless apkvol, and an answer file (see contents below).  I'm installing to the SD card of a Raspberrypi 4.

I boot the Pi from the USB stick.

I ssh into the Pi, and run "setup-alpine -f answerfile" (I removed the answer file entries to configure the disk as that was where I was suspecting an issue, I'd originally been using "syslvm").  Once the setup completed.  I rebooted.  The headless.start file is present in /etc/local.d causing /etc/network/interfaces to be overwritten.

I'm now beginning to suspect this is an RPi specific issue.  I've been digging into how the installation happens in setup-disk, but have yet to find the cause of the problem.

But I have gained a much better understanding of the Alpine boot and setup mechanisms. :-)

Answer file (blank lines and comments removed to reduce the length, and changed, hostname, domain and IPs):
```
empty_root_password=1
KEYMAPOPTS="us us"
HOSTNAMEOPTS="-n pi1"
INTERFACESOPTS="auto lo
iface lo inet loopback
auto eth0
iface eth0 inet static
	address 192.168.1.100/24
	gateway 192.168.1.1
	hostname pi1
"
DNSOPTS="-d example.com 192.168.1.9"
TIMEZONEOPTS="-z Australia/Sydney"
PROXYOPTS="none"
APKREPOSOPTS="http://mirror.aarnet.edu.au/pub/alpine/v3.14/main"
SSHDOPTS="-c openssh"
NTPOPTS="-c openntpd"
```
